internal/tofu/node_resource_abstract.go

```diff
@@ -290,33 +301,88 @@ func (n *NodeAbstractResource) DependsOn() []*addrs.Reference {
 	return result
 }
 
-func (n *NodeAbstractResource) SetProvider(p addrs.AbsProviderConfig) {
-	n.ResolvedProvider = p
+func (n *NodeAbstractResource) SetPotentialProviders(potentialProviders []distinguishableProvider) {
+	n.potentialProviders = potentialProviders
+}
+
+// resolveInstanceProvider gets the specific expanded instance of the resource, and tries to calculate the resolved
+// InstanceProvider if a few potential providers exist. A few potential providers might exist if the resource
+// configuration, or one of the containing module's configuration, contains a for_each on the providers.
+func (n *NodeAbstractResource) resolveInstanceProvider(instance addrs.AbsResourceInstance) addrs.AbsProviderConfig {
+	for _, potentialProvider := range n.potentialProviders {
+		if potentialProvider.IsResourceInstanceMatching(instance) {
+			return potentialProvider.concreteProvider.ProviderAddr()
+		}
+	}
+
+	// Can happen if ResolvedResourceProvider is already set for the whole resource, and we don't have a specific
+	// provider set for the instances
+	return addrs.AbsProviderConfig{}
```

Comment:
should we use a boolean as a second return arg here to better communicate "failure"?

Reply:
We won't use this boolean in the calling functions, so I don't think we need it